src/core/reference/src/op/loop.cpp

```diff
@@ -137,7 +139,8 @@ void loop(const std::shared_ptr<Model>& func,
 
             // Evaluate body
             body_outputs.clear();
-            reference::function(func, inputs_to_body, body_outputs);
+            body_outputs.resize(func->get_results().size());
```

Comment:
Can function reference be expanded by new implementation which accept the context.
In general the reference impl should not call `evalute` but `evalutes` should use references. 
It looks like it was don like this before and this dependency is little mixed.

But  maybe will be easier to change if there will be new `reference::function` which accept context and can be re-use in other places if required.

Reply:
ok, I've added a new `reference::function`

But as I can see `reference::function` also uses `function->evaluate` (Model::evaluate) inside.
And `Model::evaluate` uses evaluate methods of other nodes.